src/main/kotlin/net/revanced/patcher/util/Io.kt

```diff
@@ -3,47 +3,90 @@ package net.revanced.patcher.util
-                    ClassReader(jar.readBytes()).accept(classNode, ClassReader.EXPAND_FRAMES)
-                    this.add(classNode)
-                }
-                jar.closeEntry()
+import java.util.zip.ZipEntry
+import java.util.zip.ZipInputStream
+import java.util.zip.ZipOutputStream
+
+internal class Io(
+    private val input: InputStream,
+    private val output: OutputStream,
+    private val classes: MutableList<ClassNode>
+    )
+{
+    private val bufferedInputStream = BufferedInputStream(input)
+
+    fun readFromJar() {
+        bufferedInputStream.mark(0)
+       // create a BufferedInputStream in order to read the input stream again when calling saveAsJar(..)
+        val jis = JarInputStream(bufferedInputStream)
+
+        // read all entries from the input stream
+        // we use JarEntry because we only read .class files
+        var jarEntry: JarEntry?
+        while (jis.nextJarEntry.also { jarEntry = it } != null) {
+            // if the current entry ends with .class (indicating a java class file), add it to our list of classes to return
+            if (jarEntry!!.name.endsWith(".class")) {
+                // create a new ClassNode
+                val classNode = ClassNode()
+                // read the bytes with a ClassReader into the ClassNode
+                ClassReader(jis.readBytes()).accept(classNode, ClassReader.EXPAND_FRAMES)
+                // add it to our list
+                classes.add(classNode)
             }
+
+            // finally, close the entry
+            jis.closeEntry()
+        }
+
+        // at last reset the buffered input stream
+        bufferedInputStream.reset()
     }
 
-    fun writeClassesToJar(input: InputStream, output: OutputStream, classes: List<ClassNode>) {
-        val jis = JarInputStream(input)
-        val jos = JarOutputStream(output)
-
-        // TODO: Add support for adding new/custom classes
-        while (true) {
-            val next = jis.nextJarEntry ?: break
-            val e = JarEntry(next) // clone it, to not modify the input (if possible)
-            jos.putNextEntry(e)
-
-            val clazz = classes.singleOrNull {
-                clazz -> clazz.name+".class" == e.name // clazz.name is the class name only while e.name is the full filename with extension
-            };
-            if (clazz != null) {
-                val cw = ClassWriter(ClassWriter.COMPUTE_MAXS or ClassWriter.COMPUTE_FRAMES)
-                clazz.accept(cw)
-                jos.write(cw.toByteArray())
-            } else {
-                jos.write(jis.readBytes())
-            }
+    fun saveAsJar() {
+        val jis = ZipInputStream(bufferedInputStream)
+        val jos = ZipOutputStream(output)
+
+        // first write all non .class zip entries from the original input stream to the output stream
+        // we read it first to close the input stream as fast as possible
+        var zipEntry: ZipEntry?
+        while (jis.nextEntry.also { zipEntry = it } != null) {
```

Comment:
This implementation has a flaw. There is no way to ***remove*** a class. Please look into that.

Reply:
There should be, we have a List of ClassNodes, if we want to remove a class, we simply get rid it from there.

Reply:
My bad, didn't read the comment. "Non class entries", this is fine.

Reply:
Not that I know of. So this isn't an issue for now, you may want to put a FIXME or TODO above it though, so we know it's a "known issue" if it's ever going to become an issue.